cloudnet_api_client/client.py

```diff
@@ -50,48 +53,73 @@ def __init__(
-        else:
-            res = self._get_response("sites", {"type": type})
+        res = self._get_response("sites", {"type": type})
         return _build_objects(res, Site)
 
+    def site(self, site_id: str) -> Site:
+        res = self._get_response(f"sites/{site_id}")[0]
+        return _build_object(res, Site)
+
     def products(
         self, type: PRODUCT_TYPE | list[PRODUCT_TYPE] | None = None
     ) -> list[Product]:
-        res = self._get_response("products")
-        data = _build_objects(res, Product)
+        data = self._get_response("products")
         if isinstance(type, str):
-            data = [obj for obj in data if type in obj.type]
+            data = [obj for obj in data if type in obj["type"]]
         elif isinstance(type, list):
-            data = [obj for obj in data if any(t in obj.type for t in type)]
-        return data
+            data = [obj for obj in data if any(t in obj["type"] for t in type)]
+        return _build_objects(data, Product)
+
+    def product(self, product_id: str) -> ExtendedProduct:
+        res = self._get_response(f"products/{product_id}")[0]
+        obj = _build_object(res, Product)
+        return ExtendedProduct(
+            **asdict(obj),
+            derived_product_ids=_set_of_ids(res, "derivedProducts"),
+            source_instrument_ids=_set_of_ids(res, "sourceInstruments"),
+            source_product_ids=_set_of_ids(res, "sourceProducts"),
+        )
 
     def instruments(self) -> list[Instrument]:
         res = self._get_response("instrument-pids")
-        return [
-            Instrument(
-                instrument_id=obj["instrument"]["id"],
-                model=obj["model"],
-                type=obj["type"],
-                uuid=UUID(obj["uuid"]),
-                pid=obj["pid"],
-                owners=obj["owners"],
-                serial_number=obj["serialNumber"],
-                name=obj["name"],
-            )
-            for obj in res
-        ]
+        return [_create_instrument_object(obj) for obj in res]
+
+    def instrument(self, uuid: str | UUID) -> ExtendedInstrument:
+        res = self._get_response(f"instrument-pids/{uuid}")[0]
+        obj = _create_instrument_object(res)
+        return ExtendedInstrument(
+            **asdict(obj),
+            derived_product_ids=self.get_instrument_derived_products(obj.instrument_id),
+        )
+
+    def get_instrument_derived_products(self, instrument_id: str) -> frozenset[str]:
+        res = self._get_response(f"instruments/{instrument_id}")[0]
+        return _set_of_ids(res, "derivedProducts")
+
+    def models(self) -> list[Model]:
+        res = self._get_response("models")
+        return [_create_model_object(obj) for obj in res]
+
+    def model(self, model_id: str) -> Model:
+        res = self._get_response("models")
+        model = [r for r in res if r["id"] == model_id]
+        if not model:
+            raise ValueError(f"Model with id {model_id} not found")
```

Comment:
Miten virheet pitäisi käsitellä? Esim. `client.instrument('asd')` antaa `HTTPError`:in. Voisi lisätä testin virhetilanteille.